### ISSUE TITLE
2.4.11 Focus appearance (min) - complex requirement makes it difficult to write test instructions

### ISSUE THREAD

**Issue by LiLoDavis** [w3c/wcag#1460]
I've been struggling to write a manual test procedure for verifying compliance with the four bullet points. I urge you to simplify the requirement. 

The complete process involves
- 3 area calculations
  - 1 CSS pixel border of the focused control
  - a thickness of at least 8 CSS pixels along the shortest side of the element
  - a thickness of at least 2 CSS pixels (applies only to adjacent contrast)

- that are dependent on results of 3 contrast measurements
  - focus indicator color(s) vs. unfocused pixel color(s)
  - focus indicator color(s) vs. inside adjacent pixel color(s)
  - focus indicator color(s) vs. outside adjacent pixel color(s)

**Comment by mbgower** [comment]
**Draft AGWG response**
Thanks for your comment. First, the test procedures are provided in each sufficient technique for a success criterion, not in the criterion itself.

There are 3 techniques listed in the Understanding document in the current draft of the standard. Reviewing the tests for these shows that the test requirement can change dramatically depending on how the SC was technically addressed.
[G195: Using an author-supplied, highly visible focus indicator](https://www.w3.org/WAI/WCAG22/Techniques/general/G195) provides a manual 6-step test process flow:
1. Place focus on each focusable user interface element on the page using the mouse.
2. Check that there is a highly visible focus indicator.
3. Place focus on each focusable user interface element on the page using the keyboard.
4. Check that the focus indicator area is at least the size of a 1 CSS px border around the component.
5. Check that the change of contrast of the indicator between focused and unfocused states has a ratio of 3:1 for the minimum focus indicator area.
6. If the focus indicator does not have 3:1 contrast ratio with its adjacent colors, check that it is at least 2px thick.

(A [PR](https://github.com/w3c/wcag/pull/1473) has been created to remove the first two mouse steps and add an additional step for the 8px width , reducing this to a 5-step process.)
This is a fully manual process, which as written in the PR, addresses the requirements of the draft SC.

However, if you look at the test for [Creating a two-color focus indicator to ensure sufficient contrast with all components](https://www.w3.org/WAI/WCAG22/Techniques/css/C40), you'll see that it is only a 2-step process to verify, since the technical implementation of the focus indicator virtually guarantees contrast.
1. Check that the two-colors in the focus indicator are adjacent and have a contrast ratio that is 3:1 or greater with each other.
1. Check that the colors used contrast with the adjacent colors, or are at least 2 CSS pixels thick.

[to be continued]

**Comment by LiLoDavis** [comment]
Thank you for the response. Could you explain how users are expected to measure the area or thickness of a focus indicator?

**Comment by mbgower** [comment]
Hi @LiLoDavis, still working on the draft response, which then will get reviewed by the working group. However, I believe the additional material I just added may provide some guidance.

**Comment by LiLoDavis** [comment]
Thank you @mbgower. This is very helpful!

**Comment by LiLoDavis** [comment]
Could you please clarify whether "or has a thickness of at least 8 CSS pixels along the shortest side of the element" is intended to allow specifically a focus indicator that is 8 pixels thick and placed along one of the shortest sides of the element? Or is it intended to provide an alternative method of calculating the minimum area (i.e., 8px multiplied by the height or width of the control, whichever is shorter)?

**Comment by mraccess77** [comment]
My understanding is that it is meant to calculate the size of the alternative indicator but not to prescribe where it is located.

**Comment by LiLoDavis** [comment]
Thank you @mraccess77. Does "size" mean specific height and width dimensions, or the area that they define?

**Comment by alastc** [comment]
Hi @LiLoDavis,

I think it's helpful to define the size measure on the perimeter as then an outline is definitely bigger than the minimum. It is more flexible of different approaches.

**Comment by alastc** [comment]
Hi everyone,

Based on an update which was surveyed/discussed in the last two weeks, we've updated the size metric to read:

<hr>

**Minimum area**: The area of the focus indicator is either:

- at least as large as the area of a 1 CSS pixel thick perimeter of the unfocused control;
- at least as large as a 4 CSS pixels border along the shortest side, and no thinner than 2 CSS pixels.

<hr>

I'm afraid that doesn't simplify the process of measuring, but as a manual process I would approach it as: 

- Is the focus indicator area at least 2px * the perimeter?
  - If yes, is the change of contrast > 3:1?
  - Pass
- Else, is the focus indicator area at least 1px * the perimeter?
  - If yes, is the change of contrast > 3:1?
  - If yes, is the adjacent contrast > 3:1?
  - Pass
- Else, if the focus indicator area at least 4px * shortest side, and at least 2px thick.
  - If yes, is the change of contrast > 3:1?
  - If yes, is the adjacent contrast > 3:1?
  - Pass
- Else fail.

If automating it there are probably cleverer ways.

From the above thread, #1403 was merged in, and the 8px thick aspect has gone.

> how users are expected to measure the area or thickness of a focus indicator?

I wouldn't expect end-users to need this, but for designers/developers/testers, code-inspection is how I would do it.

Overall, the metrics have come from behavioural and logical standpoint, aiming to provide a minimum which is visible to users, and allows various designs to pass. That might increase the testing difficulty, but (in my opinion) I think that's the best trade-off.

I can put that to the group, or you can confirm this closes the issue for you.

**Comment by LiLoDavis** [comment]
Could you clarify whether the following minor addition (bold+italic) reflects the intent? I want to be sure that the shape of the indicator isn't being defined.

> **Minimum area**: The area of the focus indicator is either:
> 
> * at least as large as the area of a 1 CSS pixel thick perimeter of the unfocused control;
> * at least as large as **_the area of_** a 4 CSS pixels border along the shortest side, and no thinner than 2 CSS pixels.

**Comment by mraccess77** [comment]
If we allow "•	at least as large as a 4 CSS pixels border along the shortest side, and no thinner than 2 CSS pixels." does this remove the contrast minimum?   The area itself seems reasonable but if we then loose the contrast requirement I think it could be misused in ways like Detlev showed to just apply a 2px border on the bottom of a button with no contrast making it almost impossible for users to actually see the change - thus gutting a significant part of the SC.

**Comment by alastc** [comment]
> Could you clarify whether the following minor addition (bold+italic) reflects the intent? I want to be sure that the shape of the indicator isn't being defined.

These are both intended to be measures of area, not imposing a shape requirement.

> does this remove the contrast minimum?

No, if you look at the whole [SC text (now live)](https://w3c.github.io/wcag/guidelines/22/#focus-appearance-minimum) the contrast bullets are unaffected.

In practice, if it is at least 2px wide then it will automatically meet that part of 'adjacent contrast', however, I don't think we should change that bullet.

Also, [Detlev's example doesn't pass](https://github.com/w3c/wcag/pull/1594#issuecomment-767095122) the size requirement now either.

**Comment by LiLoDavis** [comment]
Thank you, @alastc, for the clarification. One more, please? 

Is there any difference in meaning between "The pixels _making up_ the minimum focus indicator area" (in **Change of contrast** bullet) and "The pixels _in_ the minimum focus indicator area" (in **Adjacent contrast area** bullet)? I'm guessing not, but the fact that they are worded differently makes me wonder.

**Comment by mraccess77** [comment]
I believe it will be very common to have short buttons and links with 16px high sides which means that the area of the indicator would be 64px minimum.  Applying a 2px bottom border with low contrast to a button that is 32px wide minimum allows the adjacent contrast clause to work with no contrast requirement thus defeating the purpose of the criteria.

**Comment by alastc** [comment]
The change of contrast bullet still applies.

**Comment by LiLoDavis** [comment]
For adjacent contrast, is it necessary to test all colors adjacent to the focus indicator? For example if a 1px outline is used to indicate focus, and the outline abuts the control (no space between the control and the indicator), is it necessary to test both the color of the control (on one side of the indicator) and the color of the background (on the other side of the indicator)?

**Comment by LiLoDavis** [comment]
And yet another question. Earlier in this thread, @alastc said:

> I'm afraid that doesn't simplify the process of measuring, but as a manual process I would approach it as:
> 
> * Is the focus indicator area at least 2px * the perimeter?
>   
>   * If yes, is the change of contrast > 3:1?
>   * Pass
> * Else, is the focus indicator area at least 1px * the perimeter?
>   
>   * If yes, is the change of contrast > 3:1?
>   * If yes, is the adjacent contrast > 3:1?
>   * Pass
> * Else, if the focus indicator area at least 4px * shortest side, and at least 2px thick.
>   
>   * If yes, is the change of contrast > 3:1?
>   * If yes, is the adjacent contrast > 3:1?
>   * Pass
> * Else fail.

For the last case, given that the focus indicator is at least 2px thick, my reading of the SC is that it is not necessary to test the adjacent contrast. Do you agree?

**Comment by mraccess77** [comment]
> For the last case, given that the focus indicator is at least 2px thick, my reading of the SC is that it is not necessary to test the adjacent contrast. Do you agree?

That's my understanding.  When it's a perimeter this is less of an issue as it makes a marked change in size  - but when this is only the bottom of a control then this is harder to detect by people with low vision unless it's a hue difference that stands out.

**Comment by alastc** [comment]
> For adjacent contrast, is it necessary to test all colors adjacent to the focus indicator?

Yes, that is the meaning we've had through various contrast guidelines, and the intent here. If a dark line is adjacent to a dark button, it is harder to see. There is a little flexibility, as outlined in the [understanding document](https://w3c.github.io/wcag/understanding/focus-appearance-minimum.html#unusual-shapes) where it is a greater size than the minimum.

> For the last case, given that the focus indicator is at least 2px thick, my reading of the SC is that it is not necessary to test the adjacent contrast. Do you agree?

Yes, sorry, I probably cut & paste that.

> Is there any difference in meaning between "The pixels making up the minimum focus indicator area" (in Change of contrast bullet) and "The pixels in the minimum focus indicator area" (in Adjacent contrast area bullet)?

No, that's just a sign of many edits being made and it needs a little clean up. I'll adjust that to "in".

Going back to the original topic, Michael posted a draft response, to which I would just add that: We came to these factors based on what is visible. There is a balance with what is testable, but I don't think there is a worthwhile change to make in adjusting that balance at the moment.

Is there anything else to clear up, or can I close this issue?

**Comment by LiLoDavis** [comment]
> Is there anything else to clear up, or can I close this issue?

I'm content, thanks.
